### PR TITLE
Maint: Improve how we capture and re-raise errors in GUI tests

### PR DESCRIPTION
**Pull request by kitchoi**
This PR improves the context manager we used for capturing exceptions that occur in the GUI event loop as a test is run.

- Change `print` to logging
- Instead of re-raising the first exception as is, serialize all the exceptions and re-raise RuntimeError instead. 
    - It is problematic to re-raise the first one regardless of its type, e.g. it could be a `SystemExit` or it could be a `StopIteration` (see [PEP 479](https://www.python.org/dev/peps/pep-0479/)). When we re-raise, we meant to make the test error, rather than propagating the intended side effect of the captured exception.
    - If more than one exception occurs, we get to see all of them.

### CODE REVIEW

File: traitsui/tests/test__tools.py [hunk 1]
```diff
@@ -179,3 +180,29 @@ def cleanup(wx_handler):
...
+        with self.assertRaises(RuntimeError) as exception_context, \
+                self.assertLogs("traitsui") as watcher:
+            with reraise_exceptions():
+                gui.invoke_later(raise_error)
+                gui.process_events()
+
+        error_msg = str(exception_context.exception)
+        self.assertIn("ZeroDivisionError", error_msg)
+        log_content, = watcher.output
+        self.assertIn("ZeroDivisionError", log_content)
```
Comment (kitchoi) [inline traitsui/tests/test__tools.py:208]: I haven't tested the handler for exceptions from trait change notification. I should do that.

**Comment by mdickinson** [comment]
I haven't looked at the code hard, but this seems like the right way to go. I especially agree with not re-raising whatever happened to be caught.

File: traitsui/tests/_tools.py [hunk 2]
```diff
@@ -24,48 +25,86 @@
...
+    def handler(object, name, old, new):
         type, value, tb = sys.exc_info()
-        exceptions.append(value)
-        _print_uncaught_exception(type, value, tb)
+        serialized_exceptions.append(_serialize_exception(type, value, tb))
+        logger.exception(
+            "Unexpected error occurred from change handler "
+            "(object: %r, name: %r, old: %r, new: %r).",
+            object, name, old, new,
+        )
```
Comment (kitchoi) [inline traitsui/tests/_tools.py:87]: I actually think this function does too much by including trait change exception handling... but I don't want to change the behaviour here just yet.

Comment (mdickinson) [reply traitsui/tests/_tools.py:87]: Agreed. It would be nice to be able to separate the two things.

Comment (mdickinson) [reply traitsui/tests/_tools.py:87]: (One day.)

**Comment by kitchoi** [comment]
Sorry for the late changes. Ready for review.

File: traitsui/tests/_tools.py [hunk 3]
```diff
@@ -24,48 +25,86 @@
...
+            msg += "\n".join(
+                "=== Exception (type: {}, value: {}) ===\n"
+                "{}".format(*record)
+                for record in serialized_exceptions
+            )
+            raise RuntimeError(msg)
+
+
+# Temporary alias to avoid conflicts across PRs.
+store_exceptions_on_all_threads = reraise_exceptions
```
Comment (rahulporuri) [inline traitsui/tests/_tools.py:107]: If this is only used internally (which i think is the case), we can just remove the alias once all of it's callers/users have been updated - preferably in a separate PR.

Comment (rahulporuri) [reply traitsui/tests/_tools.py:107]: for now, we could just have an open issue for this change.

Comment (kitchoi) [reply traitsui/tests/_tools.py:107]: I can do that renaming in a separate PR.
(The renaming is a drive-by too... the original name is a bit confusing: The motivation of this context manager has nothing to do with threading.)

Comment (kitchoi) [reply traitsui/tests/_tools.py:107]: Done: https://github.com/enthought/traitsui/pull/1102

**Pull request merged**